core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlockEncoding.java

```diff
@@ -76,4 +80,20 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
         // TODO: fix DictionaryBlock so that dictionaryIsCompacted can be set to true when the underlying block over-retains memory.
         return new DictionaryBlock(positionCount, dictionaryBlock, ids, false, new DictionaryId(mostSignificantBits, leastSignificantBits, sequenceId));
     }
+
+    @Override
+    public Optional<Block> replacementBlockForWrite(Block block)
+    {
+        DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+        if (!dictionaryBlock.isCompact()) {
+            return Optional.of(dictionaryBlock.compact());
+        }
+
+        if (dictionaryBlock.getUniqueIds() == dictionaryBlock.getPositionCount() && dictionaryBlock.isSequentialIds()) {
+            // ids mapping is identity
+            return Optional.of(dictionaryBlock.getDictionary());
+        }
```

Comment:
can this logic be part of `compact`?

Reply:
probably, but that would also change SPI contract slightly.
For example, `compacting` a dictionary means that no dictionary positions are unused. However, dictionary can still be a masking dictionary that shuffles positions. In order to extract base block (which can be done for shuffling, masking dictionary) you would have to do data copy.
However, if one cares about memory usage (e.g dictionary being compact) then no action needs to be done for shuffling dictionary that has no unused positions.

Anyways, I would prefer to keep it for later PR (if we decide to go this way)